doc/compiler/il/Node.md

```diff
@@ -0,0 +1,91 @@
+# Node
+
+## Description
+
+When an optimizer makes a transofrmation, a node is often transmuted from one opcode value to another. This can lead to a variety of bugs since the properties that belonged to the original node are implicitly inherited by the transmuted node. The old properties might have no meaning, or worse, a different meaning for the new transmuted node.
+
+To fix this problem, a Node is immutable, in that its opcode value cannot be modified once the Node has been created. Any transformation that requires a change in the Node's opcode value requires the allocation of a new node.
+
+This technique can also be applied to other properties of Nodes that should be consideted immutable. For example, the `intValue` on an `iconst` node.
+
+## Design Details
+
+### Immutable Properties of a Node
+
+Under an optimizer, we are allowed to transmute an expression as long as its "abstract value", or meaning,  remaines unchanged. To keep the expression's abstract value unchanged, we may need to change several things at once. In most cases, just changing the opcode alone will change the abstract value of the expression. For example consider the transmutation (for positive integer `x`):
+
+	x * 2	---->	x << 1
+
+This is represented as:
+```
+   *                   <<
+ /   \     ---->     /    \
+x     2             x      1
+```
+
+The expressions above have the same abstract value, because they mean the same thing. The transmutation shown is valid; but this is only true because several things happen at once: 
+
+* The `*` becomes a `<<`
+* The `2` is replaced by a `1`
+
+In general, there may be lots of things that have to happen at once during a transmutation: Multiple properties of the Node and of the children might have to be changed. We have to consider the whole expression not just the top level node here. Just changing the opcode value of the Node is not enough. In the example above, if the `*` is transmuted to a `<<` and its children remain the same, then the resulting expression is incorrect. `x * 2` does not mean the same as `x << 2`. We wish to avoid this temporary inconsistency since it is prone to bugs. The answer is to disallow intermediate (and incorrect) transmutations by not allowing just the node's opcode value to be altered. Instead, the node expression must be recreated, so the transmutation of the expression is done atomically.
+
+### Transmutating a Node Expression Inplace
+
+If we were to create a new node to replace an old one, we would normally need to update all of the references to the old node. For instance, suppose in the transformation
+
+	x * (2 ** n)  ---->  x << n
+
+we wanted to to replace `imul` by `ishl`, then we would normally need to update all of the references of the original `imul` node (the expression of which may be commoned) to refer to the new `ishl` node. Finding all these places can be expensive.
```

Comment:
If the purpose here is to discuss the problems of node opcode replacement without true recreation then limiting yourself to a simple subtree like the imul to ishl example is probably best because rearranging large expressions requires treetops and pinning etc. We probably just need to find a recreate where the flags do have different meanings and/or become invalid some of the treetop recreates like NULLCHK to treetop etc probably have this kind of behaviour.